README.md
# MWFeedParser — An RSS and Atom web feed parser for iOS

MWFeedParser is an Objective-C framework for downloading and parsing RSS (1.* and 2.*) and Atom web feeds. It is a very simple and clean implementation that reads the following information from a web feed:

#### Feed Information
- Title
- Link
- Summary

#### Feed Items
- Title
- Link
- Author name
- Date (the date the item was published)
- Updated date  (the date the item was updated, if available)
- Summary (brief description of item)
- Content (detailed item content, if available)
- Enclosures (i.e. podcasts, mp3, pdf, etc)
- Identifier (an item's guid/id)

If you use MWFeedParser on your iPhone/iPad app then please do let me know, I'd love to check it out :)

***Important:*** This free software is provided under the MIT licence (X11 license) with the addition of the following condition:

>This Software cannot be used to archive or collect data such as (but not
limited to) that of events, news, experiences and activities, for the 
purpose of any concept relating to diary/journal keeping.

The full licence can be found at the end of this document.


## Demo / Example App

There is an example iPhone application within the project which demonstrates how to use the parser to display the title of a feed, list all of the feed items, and display an item in more detail when tapped.


## Setting up the parser

Create parser:

	// Create feed parser and pass the URL of the feed
	NSURL *feedURL = [NSURL URLWithString:@"http://images.apple.com/main/rss/hotnews/hotnews.rss"];
	feedParser = [[MWFeedParser alloc] initWithFeedURL:feedURL];

Set delegate:

	// Delegate must conform to `MWFeedParserDelegate`
	feedParser.delegate = self;
	
Set the parsing type. Options are `ParseTypeFull`, `ParseTypeInfoOnly`, `ParseTypeItemsOnly`. Info refers to the information about the feed, such as it's title and description. Items are the individual items or stories.

	// Parse the feeds info (title, link) and all feed items
	feedParser.feedParseType = ParseTypeFull;
	
Set whether the parser should connect and download the feed data synchronously or asynchronously. Note, this only affects the download of the feed data, not the parsing operation itself.

	// Connection type
	feedParser.connectionType = ConnectionTypeSynchronously;
	
Initiate parsing:

	// Begin parsing
	[feedParser parse];
	
The parser will then download and parse the feed. If at any time you wish to stop the parsing, you can call:

	// Stop feed download / parsing
	[feedParser stopParsing];
	
The `stopParsing` method will stop the downloading and parsing of the feed immediately.


## Reading the feed data

Once parsing has been initiated, the delegate will receive the feed data as it is parsed.

	- (void)feedParserDidStart:(MWFeedParser *)parser; // Called when data has downloaded and parsing has begun
	- (void)feedParser:(MWFeedParser *)parser didParseFeedInfo:(MWFeedInfo *)info; // Provides info about the feed
	- (void)feedParser:(MWFeedParser *)parser didParseFeedItem:(MWFeedItem *)item; // Provides info about a feed item
	- (void)feedParserDidFinish:(MWFeedParser *)parser; // Parsing complete or stopped at any time by `stopParsing`
	- (void)feedParser:(MWFeedParser *)parser didFailWithError:(NSError *)error; // Parsing failed

`MWFeedInfo` and `MWFeedItem` contains properties (title, link, summary, etc.) that will hold the parsed data. View `MWFeedInfo.h` and `MWFeedItem.h` for more information.

***Important:*** There are some occasions where feeds do not contain some information, such as titles, links or summaries. Before using any data, you should check to see if that data exists:

	NSString *title = item.title ? item.title : @"[No Title]";
	NSString *link = item.link ? item.link : @"[No Link]";
	NSString *summary = item.summary ? item.summary : @"[No Summary]";

The method `feedParserDidFinish:` will only be called when the feed has successfully parsed, or has been stopped by a call to `stopParsing`. To determine whether the parsing completed successfully, or was stopped, you can call `isStopped`.

For a usage example, please see `RootViewController.m` in the demo project.


## Available data

Here is a list of the available properties for feed info and item objects:

#### MWFeedInfo
- `info.title` (`NSString`)
- `info.link` (`NSString`)
- `info.summary` (`NSString`)

#### MWFeedItem

- `item.title` (`NSString`)
- `item.link` (`NSString`)
- `item.author` (`NSString`)
- `item.date` (`NSDate`)
- `item.updated` (`NSDate`)
- `item.summary` (`NSString`)
- `item.content` (`NSString`)
- `item.enclosures` (`NSArray` of `NSDictionary` with keys `url`, `type` and `length`)
- `item.identifier` (`NSString`)


## Using the data

All properties of `MWFeedInfo` and `MWFeedItem` return the raw data as provided by the feed. This content may or may not include HTML and encoded entities. If the content does include HTML, you could display the data within a UIWebView, or you could use the provided `NSString` category (`NSString+HTML`) which will allow you to manipulate this HTML content. The methods available for your convenience are:

    // Convert HTML to Plain Text
    //  - Strips HTML tags & comments, removes extra whitespace and decodes HTML character entities.
    - (NSString *)stringByConvertingHTMLToPlainText;

    // Decode all HTML entities using GTM.
    - (NSString *)stringByDecodingHTMLEntities;

    // Encode all HTML entities using GTM.
    - (NSString *)stringByEncodingHTMLEntities;

    // Minimal unicode encoding will only cover characters from table
    // A.2.2 of http://www.w3.org/TR/xhtml1/dtds.html#a_dtd_Special_characters
    // which is what you want for a unicode encoded webpage.
    - (NSString *)stringByEncodingHTMLEntities:(BOOL)isUnicode;

    // Replace newlines with <br /> tags.
    - (NSString *)stringWithNewLinesAsBRs;

    // Remove newlines and white space from string.
    - (NSString *)stringByRemovingNewLinesAndWhitespace;

    // Wrap plain URLs in <a href="..." class="linkified">...</a>
    //  - Ignores URLs inside tags (any URL beginning with =")
    //  - HTTP & HTTPS schemes only
    //  - Only works in iOS 4+ as we use NSRegularExpression (returns self if not supported so be careful with NSMutableStrings)
    //  - Expression: (?<!=")\b((http|https):\/\/[\w\-_]+(\.[\w\-_]+)+([\w\-\.,@?^=%&amp;:/~\+#]*[\w\-\@?^=%&amp;/~\+#])?)
    //  - Adapted from http://regexlib.com/REDetails.aspx?regexp_id=96
    - (NSString *)stringByLinkifyingURLs;

An example of this would be:

    // Display item summary which contains HTML as plain text
	NSString *plainSummary = [item.summary stringByConvertingHTMLToPlainText];


## Debugging problems

If for some reason the parser doesn't seem to be working, try enabling Debug Logging in `MWFeedParser.h`. This will log error messages to the console and help you diagnose the problem. Error codes and their descriptions can be found at the top of `MWFeedParser.h`.


## Other information

MWFeedParser is not currently thread-safe.


## Adding to your project

### Method 1: Use CocoaPods

[CocoaPods](http://cocoapods.org) is great. If you are using CocoaPods ([and here's how to get started](http://guides.cocoapods.org/using/using-cocoapods.html)), simply add `pod 'MWFeedParser'` to your podfile and run `pod install`. You're good to go! Here's an example podfile:

```
platform :ios, '7'
    pod 'MWFeedParser'
```

If you are just interested in using the HTML and/or InternetDateTime categories in your app, you can just specify those in your podfile with `pod 'MWFeedParser/NSString+HTML'` or `pod 'MWFeedParser/NSDate+InternetDateTime'`.


### Method 2: Including Source Directly Into Your Project

1. Open `MWFeedParser.xcodeproj`.
2. Drag the `MWFeedParser` & `Categories` groups into your project, ensuring you check **Copy items into destination group's folder**.
3. Import `MWFeedParser.h` into your source as required.


## Outstanding and suggested features

- Demonstrate the previewing of formatted item summary/content (HTML with images, paragraphs, etc) within a `UIWebView` in demo app.
- Provide functionality to list available feeds when given the URL to a webpage with one or more web feeds associated with it.
- Support for the Media RSS extension (from Flickr, etc.)
- Support for the GeoRSS extension.
- Look into web feed icons.
- Look into supporting/detecting images in feed items.

Feel free to get in touch and suggest/vote for other features.


## Licence

Copyright (c) 2010 Michael Waterfall

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

1. The above copyright notice and this permission notice shall be included
   in all copies or substantial portions of the Software.

2. This Software cannot be used to archive or collect data such as (but not
   limited to) that of events, news, experiences and activities, for the 
   purpose of any concept relating to diary/journal keeping.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

CHANGELOG.txt
[2010-05-19: REVISION 1]

[NEW] Added `stopParsing` and `isStopped' methods
[CHANGED] Privatised core MWFeedParser properties and methods
[CHANGED] Improved error logging and handling
[CHANGED] Improved README documentation

Classes/MWFeedParser.m
//
//  MWFeedParser.m
//  MWFeedParser
//
//  Created by Michael Waterfall on 08/05/2010.
//  Copyright 2010 d3i. All rights reserved.
//

#import "MWFeedParser.h"
#import "MWFeedParser_Private.h"
#import "NSString+XMLEntities.h"

// NSXMLParser Logging
#if 0 // Set to 1 to enable XML parsing logs
#define MWXMLLog(x, ...) NSLog(x, ## __VA_ARGS__);
#else
#define MWXMLLog(x, ...)
#endif

// Implementation
@implementation MWFeedParser

// Properties
@synthesize delegate, url;
@synthesize urlConnection, asyncData, connectionType;
@synthesize feedParseType, feedParser, currentPath, currentText, currentElementAttributes, item, info;

#pragma mark -
#pragma mark NSObject

- (id)initWithFeedURL:(NSString *)feedURL {
	if (self = [super init]) {
		
		// URL
		self.url = [feedURL stringByReplacingOccurrencesOfString:@"feed://" withString:@"http://"];
		
		// Defaults
		feedParseType = ParseTypeFull;
		connectionType = ConnectionTypeSynchronously;

		// Date Formatters
		NSLocale *en_US_POSIX = [[NSLocale alloc] initWithLocaleIdentifier:@"en_US_POSIX"];
		dateFormatterRFC822 = [[NSDateFormatter alloc] init];
		dateFormatterRFC3339 = [[NSDateFormatter alloc] init];
        [dateFormatterRFC822 setLocale:en_US_POSIX];
        [dateFormatterRFC3339 setLocale:en_US_POSIX];
        [dateFormatterRFC822 setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0]];
        [dateFormatterRFC3339 setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0]];
		[en_US_POSIX release];
		
	}
	return self;
}

- (void)dealloc {
	[urlConnection release];
	[url release];
	[feedParser release];
	[dateFormatterRFC822 release];
	[dateFormatterRFC3339 release];
	[currentPath release];
	[currentText release];
	[currentElementAttributes release];
	[item release];
	[info release];
	[super dealloc];
}

#pragma mark -
#pragma mark Parsing

// Reset data variables before processing
- (void)reset {
	self.asyncData = nil;
	self.feedParser = nil;
	self.urlConnection = nil;
	feedType = FeedTypeUnknown;
	self.currentPath = @"/";
	self.currentText = [[NSMutableString alloc] init];
	self.item = nil;
	self.info = nil;
	hasEncounteredItems = NO;
	aborted = NO;
	stopped = NO;
	parsingComplete = NO;
	self.currentElementAttributes = nil;
}

// Begin downloading & parsing of feed
- (void)parse {
	
	// Checks
	if (!url || !delegate) {

		// Error
		[self failWithErrorCode:MWErrorCodeNotInitiated description:@"Delegate or URL not specified"];
		return;
		
	}
	
	// Reset
	[self reset];
	
	// Request
	NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:[NSURL URLWithString:url] 
												  cachePolicy:NSURLRequestReloadIgnoringLocalAndRemoteCacheData 
											  timeoutInterval:180];
	[request setValue:@"MWFeedParser" forHTTPHeaderField:@"User-Agent"];
	
	// Debug Log
	MWLog(@"MWFeedParser: Connecting & downloading feed data");
	
	// Connection
	if (connectionType == ConnectionTypeAsynchronously) {
		
		// Async
		urlConnection = [[NSURLConnection alloc] initWithRequest:request delegate:self];
		if (urlConnection) {
			
			// Create data
			asyncData = [[NSMutableData alloc] init];
			
		} else {
		
			// Error
			[self failWithErrorCode:MWErrorCodeConnectionFailed description:[NSString stringWithFormat:@"Asynchronous connection failed to URL: %@", url]];
			
		}
		
	} else {
	
		// Sync
		NSURLResponse *response = nil;
		NSError *error = nil;
		NSData *data = [NSURLConnection sendSynchronousRequest:request returningResponse:&response error:&error];
		if (data && !error) {
			
			// Process
			[self startParsingData:data];
			
		} else {
			
			// Error
			[self failWithErrorCode:MWErrorCodeConnectionFailed description:[NSString stringWithFormat:@"Synchronous connection failed to URL: %@", url]];
			
		}
		
	}
	
	// Cleanup
	[request release];
	
}

// Begin XML parsing
- (void)startParsingData:(NSData *)data {
	if (data && !feedParser) {
		
		// Create feed info
		MWFeedInfo *i = [[MWFeedInfo alloc] init];
		self.info = i;
		[i release];

		// Parse!
		feedParser = [[NSXMLParser alloc] initWithData:data];
		feedParser.delegate = self;
		[feedParser parse];
		
	}
}

// Stop parsing
- (void)stopParsing {
	
	// Stop
	stopped = YES;
	
	// Stop downloading
	[urlConnection cancel];
	self.urlConnection = nil;
	self.asyncData = nil;
	
	// Abort parsing
	aborted = YES;
	[feedParser abortParsing];
		
	// Debug Log
	MWLog(@"MWFeedParser: Parsing stopped");
	
	// Inform delegate of stop only if it hasn't already finished
	if (!parsingComplete) {
		if ([delegate respondsToSelector:@selector(feedParserDidFinish:)])
			[delegate feedParserDidFinish:self];
	}
	
}

// Abort parsing
- (void)abortParsing {
	
	// Abort
	aborted = YES;
	[feedParser abortParsing];	
	
	// Inform delegate of succesful finish
	if ([delegate respondsToSelector:@selector(feedParserDidFinish:)])
		[delegate feedParserDidFinish:self];
		
}

#pragma mark -
#pragma mark Error Handling

// If an error occurs, create NSError and inform delegate
- (void)failWithErrorCode:(int)code description:(NSString *)description {
	
	// Create error
	NSError *error = [NSError errorWithDomain:MWErrorDomain 
										 code:code 
									 userInfo:[NSDictionary dictionaryWithObject:description
																		  forKey:NSLocalizedDescriptionKey]];
	MWLog(@"%@", error);

	// Inform delegate
	if ([delegate respondsToSelector:@selector(feedParser:didFailWithError:)])
		[delegate feedParser:self didFailWithError:error];
	
}

#pragma mark -
#pragma mark NSURLConnection Delegate

- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
	[asyncData setLength:0];
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
	[asyncData appendData:data];
}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error {
	
	// Failed
	self.urlConnection = nil;
	self.asyncData = nil;
	
    // Error
	[self failWithErrorCode:MWErrorCodeConnectionFailed description:[error localizedDescription]];
	
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
	
	// Succeed
	MWLog(@"MWFeedParser: Connection successful... received %d bytes of data", [asyncData length]);
	
	// Parse
	if (!stopped) [self startParsingData:asyncData];
	
    // Cleanup
    self.urlConnection = nil;
    self.asyncData = nil;

}

-(NSCachedURLResponse *)connection:(NSURLConnection *)connection willCacheResponse:(NSCachedURLResponse *)cachedResponse {
	return nil; // Don't cache
}

#pragma mark -
#pragma mark XML Parsing

- (void)parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qualifiedName attributes:(NSDictionary *)attributeDict {
	MWXMLLog(@"NSXMLParser: didStartElement: %@", elementName);
	
	// Adjust path
	self.currentPath = [currentPath stringByAppendingPathComponent:elementName];
	self.currentElementAttributes = attributeDict;
	[self.currentText setString:@""];
	
	// Determine feed type
	if (feedType == FeedTypeUnknown) {
		if ([elementName isEqualToString:@"rss"]) feedType = FeedTypeRSS; 
		else if ([elementName isEqualToString:@"feed"]) feedType = FeedTypeAtom;
		return;
	}
	
	// Entering new feed element
	if (feedParseType != ParseTypeItemsOnly) {
		if ((feedType == FeedTypeRSS  && [currentPath isEqualToString:@"/rss/channel"]) ||
			(feedType == FeedTypeAtom && [currentPath isEqualToString:@"/feed"])) {
			return;
		}
	}
			
	// Entering new item element
	if ((feedType == FeedTypeRSS  && [currentPath isEqualToString:@"/rss/channel/item"]) ||
		(feedType == FeedTypeAtom && [currentPath isEqualToString:@"/feed/entry"])) {

		// Send off feed info to delegate
		if (!hasEncounteredItems) {
			hasEncounteredItems = YES;
			if (feedParseType != ParseTypeItemsOnly) { // Check whether to ignore feed info
				
				// Inform delegate
				if ([delegate respondsToSelector:@selector(feedParser:didParseFeedInfo:)])
					[delegate feedParser:self didParseFeedInfo:[[info retain] autorelease]];
				
				// Debug log
				MWLog(@"MWFeedParser: Feed info for \"%@\" successfully parsed", info.title);
				
				// Finish
				self.info = nil;
				
				// Stop parsing if only requiring meta data
				if (feedParseType == ParseTypeInfoOnly) {
					
					// Debug log
					MWLog(@"MWFeedParser: Parse type is ParseTypeInfoOnly so finishing here");
					
					// Finish
					[self abortParsing];
					
				}
				
			} else {
				
				// Ignoring feed info so debug log
				MWLog(@"MWFeedParser: Parse type is ParseTypeItemsOnly so ignoring feed info");
				
			}
		}
		
		// New item
		self.item = [[MWFeedItem alloc] init];
		return;
	}
	
}

- (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName {
	MWXMLLog(@"NSXMLParser: didEndElement: %@", elementName);
	
	// Store data
	BOOL processed = NO;
	if (currentText) {
		
		// Use
		switch (feedType) {
			case FeedTypeRSS: {
				
				// Item
				if (!processed) {
					if ([currentPath isEqualToString:@"/rss/channel/item/title"]) { if (currentText.length > 0) item.title = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/rss/channel/item/link"]) { if (currentText.length > 0) item.link = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/rss/channel/item/description"]) { if (currentText.length > 0) item.summary = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/rss/channel/item/content:encoded"]) { if (currentText.length > 0) item.content = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/rss/channel/item/pubDate"]) { if (currentText.length > 0) item.date = [self dateFromRFC822String:currentText]; processed = YES; }
				}
				
				// Info
				if (!processed && feedParseType != ParseTypeItemsOnly) {
					if ([currentPath isEqualToString:@"/rss/channel/title"]) { if (currentText.length > 0) info.title = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/rss/channel/description"]) { if (currentText.length > 0) info.summary = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/rss/channel/link"]) { if (currentText.length > 0) info.link = currentText; processed = YES; }
				}
				
				break;
			}
			case FeedTypeAtom: {
				
				// Item
				if (!processed) {
					if ([currentPath isEqualToString:@"/feed/entry/title"]) { if (currentText.length > 0) item.title = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/feed/entry/link"]) { NSString *link = [self linkFromAtomLinkAttributes:currentElementAttributes]; if (link) item.link = link; processed = YES; }
					else if ([currentPath isEqualToString:@"/feed/entry/summary"]) { if (currentText.length > 0) item.summary = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/feed/entry/content"]) { if (currentText.length > 0) item.content = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/feed/entry/published"]) { if (currentText.length > 0) item.date = [self dateFromRFC3339String:currentText]; processed = YES; }
				}
				
				// Info
				if (!processed && feedParseType != ParseTypeItemsOnly) {
					if ([currentPath isEqualToString:@"/feed/title"]) { if (currentText.length > 0) info.title = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/feed/description"]) { if (currentText.length > 0) info.summary = currentText; processed = YES; }
					else if ([currentPath isEqualToString:@"/feed/link"]) { NSString *link = [self linkFromAtomLinkAttributes:currentElementAttributes]; if (link) info.link = link; processed = YES; }
				}
				
				break;
			}
		}
	}
	
	// Adjust path
	self.currentPath = [currentPath stringByDeletingLastPathComponent];
	
	// If end of an item then tell delegate
	if (!processed) {
		if ((feedType == FeedTypeRSS && [elementName isEqualToString:@"item"]) ||
			(feedType == FeedTypeAtom && [elementName isEqualToString:@"entry"])) {
			
			// Ensure summary always contains data if available
			if (!item.summary) { item.summary = item.content; item.content = nil; }
			
			// Debug log
			MWLog(@"MWFeedParser: Feed item \"%@\" successfully parsed", item.title);
			
			// Inform delegate
			if ([delegate respondsToSelector:@selector(feedParser:didParseFeedItem:)])
				[delegate feedParser:self didParseFeedItem:[[item retain] autorelease]];
			
			// Finish
			self.item = nil;			
			
		}
	}
	
}

//- (void)parser:(NSXMLParser *)parser foundAttributeDeclarationWithName:(NSString *)attributeName forElement:(NSString *)elementName type:(NSString *)type defaultValue:(NSString *)defaultValue {
//	MWXMLLog(@"NSXMLParser: foundAttributeDeclarationWithName: %@", attributeName);
//}

- (void)parser:(NSXMLParser *)parser foundCDATA:(NSData *)CDATABlock {
	MWXMLLog(@"NSXMLParser: foundCDATA (%d bytes)", CDATABlock.length);
	
	// Remember characters
	NSString *string = nil;
	@try {
		
		// Try decoding with NSUTF8StringEncoding & NSISOLatin1StringEncoding
		string = [[NSString alloc] initWithData:CDATABlock encoding:NSUTF8StringEncoding];
		if (!string) string = [[NSString alloc] initWithData:CDATABlock encoding:NSISOLatin1StringEncoding];
		if (string) [currentText appendString:string];
		
	} @catch (NSException * e) { 
	} @finally {
		[string release];
	}
	
}

- (void)parser:(NSXMLParser *)parser foundCharacters:(NSString *)string {
	MWXMLLog(@"NSXMLParser: foundCharacters: %@", string);
	
	// Remember characters
	[currentText appendString:string];
	
}

- (void)parserDidStartDocument:(NSXMLParser *)parser {
	MWXMLLog(@"NSXMLParser: parserDidStartDocument");
	
	// Debug Log
	MWLog(@"MWFeedParser: Parsing started");
	
	// Inform delegate
	if ([delegate respondsToSelector:@selector(feedParserDidStart:)])
		[delegate feedParserDidStart:self];
	
}

- (void)parserDidEndDocument:(NSXMLParser *)parser {
	MWXMLLog(@"NSXMLParser: parserDidEndDocument");

	// Debug Log
	MWLog(@"MWFeedParser: Parsing finished");
	
	// Inform delegate
	parsingComplete = YES;
	if ([delegate respondsToSelector:@selector(feedParserDidFinish:)])
		[delegate feedParserDidFinish:self];
	
}

// Call if parsing error occured or parse was aborted
- (void)parser:(NSXMLParser *)parser parseErrorOccurred:(NSError *)parseError {
	MWXMLLog(@"NSXMLParser: parseErrorOccurred: %@", parseError);
	parsingComplete = YES;
	if (!aborted) {
		
		// Fail with error
		[self failWithErrorCode:MWErrorCodeFeedParsingError description:[parseError localizedDescription]];
		
	}
}

- (void)parser:(NSXMLParser *)parser validationErrorOccurred:(NSError *)validError {
	MWXMLLog(@"NSXMLParser: validationErrorOccurred: %@", validError);
	parsingComplete = YES;
	
	// Fail with error
	[self failWithErrorCode:MWErrorCodeFeedValidationError description:[validError localizedDescription]];
	
}

#pragma mark -
#pragma mark Helpers

- (NSString *)url {
	return [NSString stringWithString:url];
}

- (BOOL)isStopped {
	return stopped;
}

#pragma mark -
#pragma mark Misc

// Determine whether to use the link from atom feed (where rel is not set to "self" etc...)
- (NSString *)linkFromAtomLinkAttributes:(NSDictionary *)attributes {
	if (attributes && [attributes objectForKey:@"rel"] && [[attributes objectForKey:@"rel"] isEqualToString:@"alternate"]) {
		return [attributes objectForKey:@"href"];
	}
	return nil;
}

- (NSDate *)dateFromRFC822String:(NSString *)dateString {
	NSDate *date = nil;
	if (!date) { // Sun, 19 May 02 15:21:36 GMT
		[dateFormatterRFC822 setDateFormat:@"EEE, d MMM yy HH:mm:ss zzz"]; 
		date = [dateFormatterRFC822 dateFromString:dateString];
	}
	if (!date) { // Sun, 19 May 2002 15:21:36 GMT
		[dateFormatterRFC822 setDateFormat:@"EEE, d MMM yyyy HH:mm:ss zzz"]; 
		date = [dateFormatterRFC822 dateFromString:dateString];
	}
	if (!date) {  // Sun, 19 May 2002 15:21 GMT
		[dateFormatterRFC822 setDateFormat:@"EEE, d MMM yyyy HH:mm zzz"]; 
		date = [dateFormatterRFC822 dateFromString:dateString];
	}
	if (!date) { // Failed so Debug log
		MWLog(@"MWFeedParser: Could not parse RFC822 date: \"%@\" Possibly invalid format.", dateString);
	}
	return date;
}

- (NSDate *)dateFromRFC3339String:(NSString *)dateString {
	NSDate *date = nil;
	dateString = [dateString stringByReplacingOccurrencesOfString:@"Z" withString:@"-0000"];
	if (!date) { // 1996-12-19T16:39:57-08:00
		[dateFormatterRFC3339 setDateFormat:@"yyyy'-'MM'-'dd'T'HH':'mm':'ssZZZ"]; 
		date = [dateFormatterRFC3339 dateFromString:dateString];
	}
	if (!date) { // 1937-01-01T12:00:27.87+00:20
		[dateFormatterRFC3339 setDateFormat:@"yyyy'-'MM'-'dd'T'HH':'mm':'ss.SSSZZZ"]; 
		date = [dateFormatterRFC3339 dateFromString:dateString];
	}
	if (!date) { // Failed so Debug log
		MWLog(@"MWFeedParser: Could not parse RFC3339 date: \"%@\" Possibly invalid format.", dateString);
	}
	return date;
}

@end